src/Data/Text/Builder/Linear/Dec/Unbounded.hs

```diff
@@ -0,0 +1,380 @@
+-- • For 'BigNat's, use a buffer with `unsafePrependUnboundedDec`, then copy it.
+--
+-- For *bounded* integers we used the exact size of the decimal representation to
+-- compute the offset from which we can use the prepend action to actually append.
+--
+-- But the exact size of an (unbounded) 'Integer' could be expensive to compute.
+-- So it is faster to use a buffer and then copy it.
+unsafeAppendDec
+  ∷ ∀ s x
+   . Integer
+  → ((A.MArray s → Int → ST s Int) → ST s x)
+  → ((A.MArray s → Int → ST s Int) → ST s x)
+  → ST s x
+unsafeAppendDec n = case n of
+  I.IS i# → \append _ → append (\marr off → Bounded.unsafeAppendDec marr off (I# i#))
+  _ → \_ prepend → prepend (\marr off → unsafePrependDec marr off n)
+{-# INLINEABLE unsafeAppendDec #-}
+
+--------------------------------------------------------------------------------
+-- Prepend
+--------------------------------------------------------------------------------
+
+-- | Prepend the decimal representation of an /unbounded/ integral number.
+($$<|) ∷ Integral a ⇒ a → Buffer ⊸ Buffer
+
+infixr 6 $$<|
+n $$<| buffer = case toInteger n of
+  !n' →
+    prependBounded'
+      (maxIntegerDecLen n')
+      (\dst dstOff → unsafePrependDec dst dstOff n')
+      buffer
+{-# INLINEABLE ($$<|) #-}
+
+unsafePrependDec ∷ ∀ s. A.MArray s → Int → Integer → ST s Int
+unsafePrependDec marr off@(I# off#) n = case n of
+  I.IS i# → Bounded.unsafePrependDec marr off (I# i#)
+  _ → unsafePrependBigNatDec marr (off# -# 1#) (integerToBigNat# n) >>= prependSign
+    where
+      prependSign !off' =
+        if n < 0
+          then do
+            A.unsafeWrite marr (off' - 1) 0x2d -- '-'
+            pure (off - off' + 1)
+          else pure (off - off')
+{-# INLINEABLE unsafePrependDec #-}
+
+type DigitsWriter s = Int# → BN.BigNat# → ST s Int
+
+-- TODO: Remove once we choose a strategy
+data Strategy = SmallOnly | BigOnly | HugeOnly | Optimum
+
+-- Use the fastest writer depending on the BigNat size
+unsafePrependBigNatDec ∷ ∀ s. A.MArray s → DigitsWriter s
+unsafePrependBigNatDec marr !off0 !n0
+  | BN.bigNatSize n0 < bigSizeThreshold = prependSmallNat marr off0 n0
+  | BN.bigNatSize n0 < hugeSizeThreshold = prependBigNat marr off0 n0
+  | otherwise = prependHugeNat marr off0 n0
+  where
+    -- TODO: simplify once we choose a strategy
+    -- Strategy on the 'BigNat#' size in order to scale
+    -- strategy = SmallOnly
+    -- strategy = BigOnly
+    -- strategy = HugeOnly
+    strategy = Optimum
+    bigSizeThreshold, hugeSizeThreshold ∷ Word
+    !(bigSizeThreshold, hugeSizeThreshold) = case strategy of
+      SmallOnly → (maxBound, maxBound)
+      BigOnly → (minBound, maxBound)
+      HugeOnly → (minBound, minBound)
+      Optimum → (25, 400)
+
+-- Writer for “small” 'BigNat's.
+--
+-- Divide repeatedly by poweredBase.
+prependSmallNat ∷ ∀ s. A.MArray s → DigitsWriter s
+prependSmallNat marr = go
+  where
+    !(# power, poweredBase, _poweredBase² #) = selectPower (# #)
```

Comment:
Could you please add a comment explaining `(# #)` trick?

Reply:
I borrowed it from [`ghc-bignum`](https://hackage.haskell.org/package/ghc-bignum-1.3/docs/src/GHC.Num.BigNat.html#bigNatZero%23), which is not documented either. Added to [my other comment](https://github.com/Bodigrim/linear-builder/pull/20#discussion_r1613211828), maybe we should just drop it altogether. There is also [`bigNatFromAddrLE#`](https://hackage.haskell.org/package/ghc-bignum-1.3/docs/GHC-Num-BigNat.html#v:bigNatFromAddrLE-35-), but it buys us little as it will allocates each time and requires handling word size.

Reply:
I think I saw @BurningWitness using `(# #)` trick somewhere, but I do not remember what for.

Reply:
It is documented in a note: [GHC.Num.BigNat#BigNat](https://hackage.haskell.org/package/ghc-bignum-1.3/docs/src/GHC.Num.BigNat.html#BigNat). The usage here does exactly the same thing (though I can't say if it's worth it, I try not to import `GHC.Exts` in my projects).

---

My use case (definitely in `radix-tree`, at least) was a more general one: unboxed tuples are the only way to force the evaluation of a function without evaluating its result to WHNF .

Reply:
@BurningWitness thanks for the pointer, I missed the note. Hackage source display does not have a great contrast.

@Bodigrim I added a note for the `(# #)` use.